### PR TITLE
Add storage cache

### PR DESCRIPTION
**Pull request by ytsmiling**
## Motivation
Flushing updates of trials to persistent storages on every update (including parameter suggests) can be a significant bottleneck. This PR mitigates the bottleneck by introducing a wrapper class that wraps the `RDBStorage/ class and providing the lazy-sync capability.

## Description of the changes
- Add `CachedStorage`, which is a wrapper of `RDBStorage` and provide the lazy-sync capability.
    - While `CachedStorage` only supports `RDBStorage`, it is straightforward to support other classes such as `RedisStorage`. Thus, `CachedStorage` file is placed directly under the `optuna/storages` class.
- Introduce two additional public methods to `RDBStorage`.
    - One returns `FrozenTrial` when creating a new trial. Note, `create_new_trial` in `BaseStorage` only returns the trial-id.
    - The other receives a `FrozenTrial` and flush the updates to persistent storages.

## Current Status
- [x] Implement `CachedStorage` class.
- [x] Add tests.
- [x] Add docs where necessary.

## Microbench
(DB: official mysql docker image (tag:8.0.19), storage: SSD, CPU: Intel Core i7-6700K, python: 3.5.2)
- perf:optimize-study: time taken for a single optimization

### This PR
```
{'n_study': 1, 'n_trial': 100, 'n_param': 3}
perf:optimize-study                               0:00:10.169289
{'n_study': 1, 'n_trial': 100, 'n_param': 30}
perf:optimize-study                               0:00:47.566834
```

### Current master
```
{'n_study': 1, 'n_trial': 100, 'n_param': 3}
perf:optimize-study                               0:00:22.499407
{'n_study': 1, 'n_trial': 100, 'n_param': 30}
perf:optimize-study                               0:02:33.151298
```

Benchmark code:
```python
import argparse
from collections import defaultdict
from datetime import datetime
import math

import sqlalchemy
import optuna

profile_result = defaultdict(lambda: datetime.now() - datetime.now())


class Profile:
    def __init__(self, name):
        self.name = name

    def __enter__(self):
        self.start = datetime.now()

    def __exit__(self, exc_type, exc_val, exc_tb):
        profile_result[self.name] += datetime.now() - self.start


def print_profile():
    for key, value in sorted(profile_result.items(), key=lambda i: i[1],
                             reverse=True):
        print(key.ljust(50) + '{}'.format(value))


def build_objective_fun(n_param):
    def objective(trial):
        return sum([
            math.sin(trial.suggest_float('param-{}'.format(i), 0, math.pi * 2))
            for i in range(n_param)
        ])

    return objective


def define_flags(parser):
    parser.add_argument('mysql_user', type=str)
    parser.add_argument('mysql_password', type=str)
    parser.add_argument('mysql_host', type=str)
    parser.add_argument('n_study', type=int)
    parser.add_argument('n_trial', type=int)
    parser.add_argument('n_param', type=int)
    return parser


if __name__ == '__main__':
    parser = define_flags(argparse.ArgumentParser())
    args = parser.parse_args()
    print(vars(args))

    storage_str = 'mysql+pymysql://{}:{}@{}/'.format(
        args.mysql_user,
        args.mysql_password,
        args.mysql_host,
    )
    engine = sqlalchemy.create_engine(storage_str)
    sampler = optuna.samplers.TPESampler()
    conn = engine.connect()
    conn.execute("commit")
    database_str = 'profile_storage_s{}_t{}_p{}'.format(
        args.n_study, args.n_trial, args.n_param)

    try:
        conn.execute(
            "drop database {}".format(database_str))
    except:
        pass
    conn.execute("create database {}".format(
        database_str
    ))
    conn.close()

    for i in range(args.n_study):
        storage = optuna.storages.get_storage(storage_str + database_str)
        study = optuna.create_study(sampler=sampler, storage=storage)
        study_id = study.study_id
        with Profile('perf:optimize-study'):
            study.optimize(build_objective_fun(args.n_param), n_trials=args.n_trial, gc_after_trial=False)

    print_profile()

```

### CODE REVIEW

**Comment by ytsmiling** [comment]
After fixing some bugs, I updated the microbenchmark.
```
{'n_study': 1, 'n_trial': 100, 'n_param': 30}
perf:optimize-study                               0:00:45.771852
{'n_study': 1, 'n_trial': 100, 'n_param': 3}
perf:optimize-study                               0:00:13.391903
```

File: optuna/storages/cached_storage.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+        param_name: str,
+        param_value_internal: float,
+        distribution: distributions.BaseDistribution,
+    ) -> bool:
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, updates = self._hot_cache[trial_id]
+                study_id = self._trial_id_to_study_id[trial_id]
+                cached_dist = self._param_distributions[study_id].get(param_name, None)
```
Comment (sile) [inline optuna/storages/cached_storage.py:144]: Since `RDBStorage` performs the distribution compatibility check only at trial-level (see below), it seems that `CachedStorage` also can do it the same level and remove the `_param_distributions` field.
https://github.com/optuna/optuna/blob/5c07d0afffeb506866860fa2ca21a71a4199c075/optuna/storages/rdb/storage.py#L547-L549

```suggestion
                cached_dist = cached_trial.distributions.get(param_name, None)
```

File: optuna/storages/cached_storage.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+    def set_trial_system_attr(self, trial_id: int, key: str, value: Any) -> None:
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, updates = self._hot_cache[trial_id]
+                self._dirty_trials.add(trial_id)
+                cached_trial.system_attrs[key] = value
+                updates.system_attrs[key] = value
+                return
```
Comment (sile) [inline optuna/storages/cached_storage.py:239]: Since `SuccessiveHalvingPruner` accesses to the system attributes of other trials (including RUNNING ones), it seems needed to delete this block to keep compatibility with the existing behavior.
https://github.com/optuna/optuna/blob/5c07d0afffeb506866860fa2ca21a71a4199c075/optuna/pruners/successive_halving.py#L221

File: optuna/storages/rdb/storage.py [hunk 3]
```diff
@@ -437,6 +435,25 @@ def get_all_study_summaries(self) -> List[StudySummary]:
     def create_new_trial(self, study_id, template_trial=None):
         # type: (int, Optional[FrozenTrial]) -> int
 
+        return self.create_new_trial_with_trial(study_id, template_trial)[0]
+
+    def create_new_trial_with_trial(
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/rdb/storage.py:440]: In that sense, all methods should have `_` prefix.
If this class turned out to be successful, I think it's beneficial to add this method to `BaseStorage` class interface and make this cache class available for all other storage backends.
Btw, I think this method should have a better name, which I'll address later.

File: optuna/storages/rdb/storage.py [hunk 4]
```diff
@@ -437,6 +435,25 @@ def get_all_study_summaries(self) -> List[StudySummary]:
...
+        self, study_id: int, template_trial: Optional[FrozenTrial] = None
+    ) -> Tuple[int, FrozenTrial]:
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/rdb/storage.py:442]: I'd like to avoid accessing the private attribute here.

[the remark this answers is not in the document]
Comment (sile) [reply optuna/storages/rdb/storage.py:442]: Okay. I'll mark this conversation as resolved.

File: optuna/storages/rdb/storage.py [hunk 5]
```diff
@@ -488,11 +505,122 @@ def create_new_trial(self, study_id, template_trial=None):
...
+            )
+
         session.add(trial)
 
         self._commit(session)
 
-        return trial.trial_id
+        return trial.trial_id, frozen
+
+    def update_trial(
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/rdb/storage.py:534]: Do you think it's better to make this method private this time and make it public when it becomes a member of the `BaseStorage` interface? If not, I'll leave this as is.

File: optuna/storages/cached_storage.py (hunk 1 above)
Comment (ytsmiling) [reply optuna/storages/cached_storage.py:144]: The model class checks the distribution compatibility in the study level. (And both of current PR and your suggestion are incompatible with the current implementation.)

File: optuna/storages/rdb/storage.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sile) [reply optuna/storages/rdb/storage.py:440]: > all methods should have `_` prefix.

Yes, I basically think so about the methods of `RDBStorage` (it's not supposed to be accessed by the library users directly). An exception is the methods defined by `BaseStorage`, so if this method is added to the interface of the base class, we will be able to remove the `_` prefix at that timing.
We may need to update [the Coding Style Conventions](https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#add-_-prefix-to-the-names-of-private-methods-functions-fields-and-classes) doc as it doesn't mention base class methods.

> I think this method should have a better name, which I'll address later.

👍

File: optuna/storages/cached_storage.py (hunk 1 above)
Comment (ytsmiling) [reply optuna/storages/cached_storage.py:145]: [This line](https://github.com/optuna/optuna/blob/5c07d0afffeb506866860fa2ca21a71a4199c075/optuna/storages/rdb/models.py#L400) checks the study-level distribution compatibility in RDB storage.

Comment (ytsmiling) [reply optuna/storages/cached_storage.py:145]: I fixed this PR so that it checks distribution compatibility across workers.

File: optuna/storages/cached_storage.py (hunk 2 above)
Comment (ytsmiling) [reply optuna/storages/cached_storage.py]: I fixed this issue.

File: optuna/storages/cached_storage.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, updates = self._hot_cache[trial_id]
+                cached_trial.state = state
+                updates.state = state
+                if trial_id in self._dirty_trials:
+                    self._dirty_trials.remove(trial_id)
+                ret = self._update_storage(trial_id)
+                if cached_trial.state.is_finished():
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/cached_storage.py]: Applied your suggestion.

File: optuna/storages/cached_storage.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+    def create_new_trial(self, study_id: int, template_trial: Optional[FrozenTrial] = None) -> int:
+
+        trial_id, frozen_trial = self._storage.create_new_trial_with_trial(
+            study_id, template_trial
+        )
+        with self._lock:
+            self._study_id_to_trial_id[study_id].append(trial_id)
+            self._trial_id_to_study_id[trial_id] = study_id
+            self._hot_cache[trial_id] = frozen_trial, _TrialUpdates()
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/cached_storage.py]: Applied your suggestion.

File: optuna/storages/rdb/storage.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/rdb/storage.py]: Added `_` prefix, which I hope to be removed in the near future.

File: optuna/storages/rdb/storage.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/rdb/storage.py]: Added the prefix.

File: optuna/storages/__init__.py [hunk 8]
```diff
@@ -1,20 +1,22 @@
...
-from optuna.storages.redis import RedisStorage  # NOQA
-from optuna.storages.rdb.storage import RDBStorage  # NOQA
+from optuna.storages.base import BaseStorage
+from optuna.storages.cached_storage import CachedStorage as _Cache
+from optuna.storages.in_memory import InMemoryStorage
+from optuna.storages.rdb.storage import RDBStorage
+from optuna.storages.redis import RedisStorage
 
-from typing import Union  # NOQA
+from typing import Union
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/__init__.py]: Resolved this issue.

File: optuna/storages/cached_storage.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+from optuna import distributions
+from optuna.storages import base
+from optuna.storages.rdb.storage import RDBStorage
+from optuna.study import StudyDirection
+from optuna.study import StudySummary
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+class CachedStorage(base.BaseStorage):
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/cached_storage.py]: Resolved this issue.

File: optuna/storages/cached_storage.py (hunk 1 above)
Comment (sile) [reply optuna/storages/cached_storage.py:145]: I see. Thank you for the explanation!

File: optuna/storages/rdb/storage.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sile) [reply optuna/storages/rdb/storage.py]: > Do you think it's better to make this method private this time and make it public when it becomes a member of the `BaseStorage` interface?

Yes, I think so.

File: optuna/storages/cached_storage.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+                else:
+                    self._param_distributions[study_id][param_name] = distribution
+                if param_name in cached_trial.params:
+                    return False
+                self._dirty_trials.add(trial_id)
+                cached_trial.params[param_name] = distribution.to_external_repr(
+                    param_value_internal
+                )
+                cached_trial.distributions[param_name] = distribution
+                if cached_dist:
```
Comment (grafi-tt) [inline optuna/storages/cached_storage.py:157]: I guess the line

```
self._dirty_trials.add(trial_id)
```

should be moved under `if cached_dist:` .

The logic here is a bit difficult too. Some comments, like "If cached_dist isn't available, we have to access the storage in order to check distribution compatibility." would help readers.

Comment (ytsmiling) [reply optuna/storages/cached_storage.py:157]: Thank you for your comment. Indeed, the line should be moved, and the whole logic requires more readability. I'll address them.

**Comment by ytsmiling** [comment]
This PR should not be merged before #1191.

File: optuna/storages/cached_storage.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+from optuna import distributions
+from optuna.storages import base
+from optuna.storages.rdb.storage import RDBStorage
+from optuna.study import StudyDirection
+from optuna.study import StudySummary
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+class _CachedStorage(base.BaseStorage):
```
Comment (toshihikoyanase) [inline optuna/storages/cached_storage.py:21]: [Opinion Wanted] How about making this class public?
Currently, the `RDBStorage` instance created via the `get_storage` function will be wrapped by `_CachedStorage`, but users can create `RDBStorage` instance directly. It sometimes required to pads DB options such as `pool_pre_ping` and `wait_timeout` for MySQL (c.f., #622) . In such cases, users may want to apply cache to the `RDBStorage` instance by themselves.

Comment (toshihikoyanase) [reply optuna/storages/cached_storage.py:21]: Ah, `get_storage` wraps `RDBStorage` and `get_storage` is called by `create_study`. So, `RDBStorage` will be wrapped by `_CachedStorage` and we can keep it private.

**Comment by ytsmiling** [comment]
I rearranged commits and now this PR passed the updated tests #1191.
I addressed review comments and would like to ask re-reviews.

Some implementation has been changed from the latest reviews:
- CachedStorage uses `_StudyInfo`, which is similar to the current implementation of InMemoryStorage.

@sile @toshihikoyanase Please review this PR when you have time.

File: optuna/storages/cached_storage.py [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+from optuna import distributions
+from optuna.storages import base
+from optuna.storages.rdb.storage import RDBStorage
+from optuna.study import StudyDirection
+from optuna.study import StudySummary
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+class _TrialUpdates:
```
Comment (toshihikoyanase) [inline optuna/storages/cached_storage.py:18]: [Question] Could you tell me why this method name is plural? IMO, it can be `_TrialUpdate`.

File: optuna/storages/cached_storage.py [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+        self._lock = threading.RLock()
+
+    def __getstate__(self):
+        # type: () -> Dict[Any, Any]
+        state = self.__dict__.copy()
+        del state["_lock"]
+        return state
+
+    def __setstate__(self, state):
+        # type: (Dict[Any, Any]) -> None
```
Comment (sile) [inline optuna/storages/cached_storage.py:60]: Ditto:
```suggestion
    def __setstate__(self, state: Dict[Any, Any]) -> None:
```

File: optuna/storages/cached_storage.py [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+    """
+
+    def __init__(self, backend: RDBStorage) -> None:
+        self._backend = backend
+        self._studies = {}  # type: Dict[int, _StudyInfo]
+        self._trial_id_to_study_id_and_number = dict()  # type: Dict[int, Tuple[int, int]]
+        self._lock = threading.RLock()
+
+    def __getstate__(self):
+        # type: () -> Dict[Any, Any]
```
Comment (sile) [inline optuna/storages/cached_storage.py:54]: A nitpicking but let me suggest the following change:
```suggestion
    def __getstate__(self) -> Dict[Any, Any]:
```

File: optuna/storages/cached_storage.py [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+            if study_id in self._studies:
+                for key, trial in self._studies[study_id].trials.items():
+                    trials[trial.number] = trial
+            return copy.deepcopy(trials) if deepcopy else trials
+
+    def get_n_trials(self, study_id: int, state: Optional[TrialState] = None) -> int:
+
+        return self._backend.get_n_trials(study_id, state)
+
+    def flush(self) -> None:
```
Comment (sile) [inline optuna/storages/cached_storage.py:303]: It seems that this method is not used in this PR. So how about removing it at the moment?
Or, if we keep this as a public method, I think that it's needed to add `with self._lock:` at the head of the body.

File: optuna/storages/cached_storage.py [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+        self.values = dict()  # type: Dict[int, float]
+        self.user_attrs = dict()  # type: Dict[str, Any]
+        self.system_attrs = dict()  # type: Dict[str, Any]
+        self.params = dict()  # type: Dict[str, Any]
+        self.distributions = dict()  # type: Dict[str, distributions.BaseDistribution]
+
+
+class _StudyInfo:
+    def __init__(self) -> None:
+        self.trials = dict()  # type: Dict[int, FrozenTrial]
```
Comment (sile) [inline optuna/storages/cached_storage.py:31]: This is not a strong opinion but I like `running_trials` as the field name.
(you can ignore this comment)

Comment (ytsmiling) [reply optuna/storages/cached_storage.py:31]: Thank you for the comment. I agree with that, but I'm planning to include finished trials into _StudyInfo in the next PR. So let me keep the name.

File: optuna/storages/cached_storage.py (hunk 12 above)
Comment (ytsmiling) [reply optuna/storages/cached_storage.py:18]: I'll replace the name with `_TrialUpdate`.

File: optuna/storages/cached_storage.py (hunk 16 above)
Comment (sile) [reply optuna/storages/cached_storage.py:31]: I see. Thank you for your response.

File: optuna/storages/cached_storage.py (hunk 15 above)
Comment (ytsmiling) [reply optuna/storages/cached_storage.py]: Thank you. I removed the method.

File: optuna/storages/rdb/storage.py [hunk 17]
```diff
@@ -450,9 +448,28 @@ def get_all_study_summaries(self) -> List[StudySummary]:
     def create_new_trial(self, study_id, template_trial=None):
         # type: (int, Optional[FrozenTrial]) -> int
 
+        return self._create_new_trial_with_trial(study_id, template_trial)[0]
+
+    def _create_new_trial_with_trial(
+        self, study_id: int, template_trial: Optional[FrozenTrial] = None
+    ) -> Tuple[int, FrozenTrial]:
+        """Create a new trial and returns trial_id and the trial converted into a FrozenTrial.
```
Comment (toshihikoyanase) [inline optuna/storages/rdb/storage.py:456]: It is a bit difficult for me to parse this sentence. If possible, could you rephrase it? The following is just an example.
```suggestion
        """Create a new trial and return trial_id with the trial converted into a :class:`~optuna.trial.FrozenTrial`.
```

File: optuna/storages/rdb/storage.py [hunk 18]
```diff
@@ -450,9 +448,28 @@ def get_all_study_summaries(self) -> List[StudySummary]:
...
+
+        Args:
+            study_id:
+                Study id.
+            template_trial:
+                A FrozenTrial with default trial attributes.
```
Comment (toshihikoyanase) [inline optuna/storages/rdb/storage.py:462]: [nit] Please ignore if not relevant. Is the following sentence too verbose?
```suggestion
                A :class:`~optuna.trial.FrozenTrial` with default values for trial attributes.
```

File: optuna/storages/rdb/storage.py [hunk 19]
```diff
@@ -450,9 +448,28 @@ def get_all_study_summaries(self) -> List[StudySummary]:
...
+
+        Returns:
+            A tuple of the trial_id of the created trial and the trial converted
+            into a FrozenTrial instance.
```
Comment (toshihikoyanase) [inline optuna/storages/rdb/storage.py:466]: [nit] Please add an internal link to `FrozenTrial`.
```suggestion
            into a :class:`~optuna.trial.FrozenTrial` instance.
```

File: tests/storages_tests/test_cached_storage.py [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+
+                    storage.set_trial_state(trial_id, state)
+                    assert update_mock.call_count == 1
+                    assert set_mock.call_count == 0
+                    assert get_mock.call_count == 0
+
+                    storage.get_trial(trial_id)
+                    assert update_mock.call_count == 1
+                    assert set_mock.call_count == 0
+                    assert get_mock.call_count == 1
```
Comment (toshihikoyanase) [inline tests/storages_tests/test_cached_storage.py:133]: I think that this nest is a bit deep. How about concatenating with-statements by commas? (c.f., https://docs.python.org/3/reference/compound_stmts.html#the-with-statement)
If it's OK, please apply similar changes to other test cases.

```suggestion
        with patch.object(base_storage, "_update_trial", return_value=True) as update_mock, \
            patch.object(base_storage, "set_trial_state", return_value=True) as set_mock, \
            patch.object(base_storage, "get_trial", return_value=True) as get_mock:

            storage.get_trial(trial_id)
            assert update_mock.call_count == 0
            assert set_mock.call_count == 0
            assert get_mock.call_count == 0

            storage.set_trial_state(trial_id, state)
            assert update_mock.call_count == 1
            assert set_mock.call_count == 0
            assert get_mock.call_count == 0

            storage.get_trial(trial_id)
            assert update_mock.call_count == 1
            assert set_mock.call_count == 0
            assert get_mock.call_count == 1
```

File: tests/storages_tests/rdb_tests/test_storage.py [hunk 21]
```diff
@@ -215,3 +217,103 @@ def setup_trials(storage, study_id):
...
+            {"user_attrs": {"attrA": 2.3, "attrB": "foo"}},
+            {"user_attrs": {"attrA": 2.3, "attrB": "foo"}},
+        ),
+        (
+            {"system_attrs": {"attrC": 2.3, "attrB": "bar"}},
+            {"system_attrs": {"attrC": 2.3, "attrB": "bar"}},
+        ),
+    ]  # type: List[Tuple[Dict[str, Any], Dict[str, Any]]]
+
+    for fields_to_modify, kwargs in test_sets:
```
Comment (toshihikoyanase) [inline tests/storages_tests/rdb_tests/test_storage.py:262]: IMO, we can make the logic shallow if you use `@pytest.mark.parametrize`. What do you think?

```python
@pytest.mark.parametrize(
    "fields_to_modify, kwargs",
    [
        (
            {"state": TrialState.COMPLETE, "datetime_complete": None},
            {"state": TrialState.COMPLETE},
        ),
        ...
    ],
)
def test_update_trial(fields_to_modify: Dict[str, Any], kwargs: Dict[str, Any]) -> None:
    storage = create_test_storage()
    study_id = storage.create_new_study()

    trial_id = storage.create_new_trial(study_id)
    trial_before_update = storage.get_trial(trial_id)
    ....
```

Comment (ytsmiling) [reply tests/storages_tests/rdb_tests/test_storage.py:262]: It looks nice and I'll modify the code accordingly.

File: tests/storages_tests/test_cached_storage.py (hunk 20 above)
Comment (ytsmiling) [reply tests/storages_tests/test_cached_storage.py:133]: Thank you for the suggestion. I'll apply the change to all test cases.

File: optuna/storages/rdb/storage.py [hunk 22]
```diff
@@ -508,7 +525,165 @@ def create_new_trial(self, study_id, template_trial=None):
...
+        return trial.trial_id, frozen
+
+    def _update_trial(
+        self,
+        trial_id: int,
+        state: Optional[TrialState] = None,
+        value: Optional[float] = None,
+        values: Optional[Dict[int, float]] = None,
+        params: Optional[Dict[str, Any]] = None,
+        dists: Optional[Dict[str, distributions.BaseDistribution]] = None,
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/rdb/storage.py:557]: I was just too lazy, and I'll replace it with `distributions_`.
Note: The name `distributions` conflicts, and thus, I'll add the underscore.

File: optuna/storages/rdb/storage.py (hunk 17 above)
Comment (ytsmiling) [reply optuna/storages/rdb/storage.py:456]: I think we can replace it with
```Create a new trial and return trial_id with a :class:`~optuna.trial.FrozenTrial`.```

**Comment by ytsmiling** [comment]
@toshihikoyanase Thank you for reviewing this PR.
> 1. FrozenTrial is not frozen anymore. The attribute values are updated to keep the cache up-to-date. Not now, but we may rename it in the future.

Actually, the mutation of `FrozenTrial` was not introduced in this PR, but it has been a long history in the `InMemoryStorage` class.

> 2. This PR directly manipulates SQLAlchemy queries while the previous implementation tried to capslate them in optuna/rdb/models.py. It may be inevitable to speed up storages.

Yes, I think we should remove methods in the `models` package and directly manipulate `sqlalchemy`. It can result in a significant performance gain (e.g. get_all_study_summaries method). From a pure performance perspective, it's even better to stop using orm and directly use SQL, but it requires much more code changes and I think we do not need to switch to SQL for now.

File: optuna/storages/rdb/storage.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (toshihikoyanase) [reply optuna/storages/rdb/storage.py]: Thanks! I got it.

File: tests/storages_tests/test_cached_storage.py (hunk 20 above)
Comment (toshihikoyanase) [reply tests/storages_tests/test_cached_storage.py]: Thank you for your update!

File: tests/storages_tests/rdb_tests/test_storage.py (hunk 21 above)
Comment (toshihikoyanase) [reply tests/storages_tests/rdb_tests/test_storage.py]: Thanks! I confirmed the change https://github.com/optuna/optuna/pull/1140/commits/740c5961116aa1a6ae8a1637c8a87b9c62a1bc55.

File: optuna/storages/rdb/storage.py (hunk 17 above)
Comment (toshihikoyanase) [reply optuna/storages/rdb/storage.py]: I confirmed the change in  https://github.com/optuna/optuna/pull/1140/commits/5ae39047691fcb01e1a6ac8956049ab76beb5a13.
It looks simpler and easier to understand. Thanks!
